### PR TITLE
Add form state hook for VAOS form pages

### PR DESCRIPTION
**Pull request by jbalboni**
## Description
Currently we hold form system state (schema, uiSchema, and data) in Redux. However, we never really use schemas across different pages, they're really just used on their specific page. So, this is an experiment to manage that form state in a hook, using the existing form system functions.

We should go over this as a team and see what we all think.

This PR
- Converts the TypeOfCarePage and the ContactInfoPage to use useFormState instead of putting form schemas in Redux
- Removes unused Redux code

## Testing done
Local and unit testing

## Acceptance criteria
- [ ] Things work and team things this is a better approach

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/applications/vaos/hooks/useFormState.jsx [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ *   hook is called
+ * @param {Array} [params.dependencies=[]] Array of useEffect dependencies that will force the form state to be reset using the current
+ *   values for initialSchema and uiSchema.
+ * @param {boolean} [params.enabled=true] Indicates if the form state is enabled or not. All values will be returned as null
+ *   when this is false, and when it changes to true the form state will be set up using the current schema and uiSchema
+ *   values
+ * }
+ * @returns {{ schema: Object, uiSchema: Object, data: Object }} The form state values, to be used with SchemaForm
+ */
+export default function useFormState({
```
Comment (jbalboni) [inline src/applications/vaos/hooks/useFormState.jsx:42]: This hook might be hard to follow, but it reuses the form system functions we call in our Redux actions to set up and change form state. In addition it provides some mechanisms for making sure we don't run our initial schema update code over and over (`initialSchema` can be a function) and to allow us to update our form after data changes later (like after an api call) with `dependencies` and `enabled`.

File: src/applications/vaos/new-appointment/components/ContactInfoPage.jsx [hunk 2]
```diff
@@ -91,21 +100,27 @@ const uiSchema = {
...
+  const userData = useSelector(getFormData);
+  const email = useSelector(selectVAPEmailAddress);
+  const homePhone = useSelector(selectVAPHomePhoneString);
+  const mobilePhone = useSelector(selectVAPMobilePhoneString);
+  const { data, schema, setData } = useFormState({
+    initialSchema,
+    uiSchema,
+    initialData: {
+      ...userData,
+      email: userData.email || email,
```
Comment (jbalboni) [inline src/applications/vaos/new-appointment/components/ContactInfoPage.jsx:120]: This pre-population logic was in the type of care action/reducer, but now it can be here, and it's arguably clearer.

File: src/applications/vaos/new-appointment/components/ContactInfoPage.jsx [hunk 3]
```diff
@@ -91,21 +100,27 @@ const uiSchema = {
...
-}) {
-  const history = useHistory();
+export default function ContactInfoPage() {
   useEffect(() => {
-    openFormPage(pageKey, uiSchema, initialSchema);
     document.title = `${pageTitle} | Veterans Affairs`;
     scrollAndFocus();
   }, []);
+  const history = useHistory();
+  const dispatch = useDispatch();
```
Comment (jbalboni) [inline src/applications/vaos/new-appointment/components/ContactInfoPage.jsx:109]: I also ended up removing `connect()` in here, since it was easier to use the React-Redux hooks.

File: src/applications/vaos/new-appointment/components/TypeOfCarePage/index.jsx [hunk 4]
```diff
@@ -70,6 +57,35 @@ function TypeOfCarePage({
     }
   }, []);
 
+  const { data, schema, setData, uiSchema } = useFormState({
+    initialSchema: () => {
```
Comment (jbalboni) [inline src/applications/vaos/new-appointment/components/TypeOfCarePage/index.jsx:61]: This logic was in the reducer before, but now it's inline with the component and you don't have to jump between files (or even back up to the schema variables) to see what's happening.

File: src/applications/vaos/new-appointment/components/TypeOfCarePage/index.jsx [hunk 5]
```diff
@@ -92,7 +108,7 @@ function TypeOfCarePage({
           title="Type of care"
           schema={schema}
           uiSchema={uiSchema}
-          onSubmit={() => routeToNextAppointmentPage(history, pageKey)}
+          onSubmit={() => routeToNextAppointmentPage(history, pageKey, data)}
```
Comment (jbalboni) [inline src/applications/vaos/new-appointment/components/TypeOfCarePage/index.jsx:111]: One notable change is that we need to save the form data whenever we change pages, because any user updates are held in component state by default now.

File: src/applications/vaos/new-appointment/redux/reducer.js [hunk 6]
```diff
@@ -245,10 +222,32 @@ export default function formReducer(state = initialState, action) {
           [action.pageKey]: 'home',
         };
       }
+      let newPages = state.pages;
```
Comment (jbalboni) [inline src/applications/vaos/new-appointment/redux/reducer.js:225]: I moved this from the other reducer because with the hook, we aren't using that form data update action, but everything is using this action still.

File: src/applications/vaos/new-appointment/redux/reducer.js [hunk 7]
```diff
@@ -245,10 +222,32 @@ export default function formReducer(state = initialState, action) {
...
+      ) {
+        newPages = unset('ccPreferences', newPages);
+        newData = set('communityCareProvider', {}, newData);
+      }
+
       return {
         ...state,
         pageChangeInProgress: true,
         previousPages: updatedPreviousPages,
+        data: newData,
```
Comment (jbalboni) [inline src/applications/vaos/new-appointment/redux/reducer.js:249]: Update the form data here every time the page changes.

Also maybe worth noting that using the back and forward buttons will not persist your form data changes, which I don't think is necessarily bad. We could account for that if we need to, though.

**Comment by LanceGundersen** [comment]
I am a fan of this. Matches the conventions of other frameworks and is easier (in my opinion) to grok.

File: src/applications/vaos/hooks/useFormState.jsx [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+}) {
+  const initialLoadRef = useRef(true);
+  const dataUpdatedRef = useRef(false);
+  const [formState, setFormState] = useState(() => {
+    let schema = initialSchema;
+    if (typeof initialSchema === 'function') {
+      schema = initialSchema();
+    }
+
+    if (!schema) {
```
Comment (jbalboni) [inline src/applications/vaos/hooks/useFormState.jsx:53]: Basically, passing a falsy schema will be like disabling the form state calculations

File: src/applications/vaos/hooks/useFormState.jsx [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+  useEffect(() => {
+    if (dependencies?.length && !initialLoadRef.current) {
+      let schema = initialSchema;
+      if (typeof initialSchema === 'function') {
+        schema = initialSchema();
+      }
+      setFormState(
+        setupFormData(
+          dataUpdatedRef.current ? formState.data : initialData,
```
Comment (jbalboni) [inline src/applications/vaos/hooks/useFormState.jsx:68]: I was trying to handle cases where data might get updated through a Redux action after the first render. I could have a useEffect that looked for `initialData` changes, but that puts someone at risk of creating an infinite loop if they happen to pass something to `initialData` that has a different reference each time.

**Review by vbahinwillit — APPROVED** [review]
